### PR TITLE
feat: return most backend API errors to frontend

### PR DESCRIPTION
**Pull request by yehong-z**
Fix: https://github.com/casdoor/casdoor/issues/1831

### CODE REVIEW

File: object/organization.go [hunk 1]
```diff
@@ -134,10 +134,10 @@ func GetMaskedOrganizations(organizations []*Organization) []*Organization {
 	return organizations
 }
 
-func UpdateOrganization(id string, organization *Organization) bool {
+func UpdateOrganization(id string, organization *Organization) (bool, error) {
 	owner, name := util.GetOwnerAndNameFromId(id)
 	if getOrganization(owner, name) == nil {
-		return false
+		return false, nil
```
Comment (hsluoyz) [inline object/organization.go:140]: return i18n error: `The object does not exist: org_id`

File: controllers/organization.go [hunk 2]
```diff
@@ -79,8 +79,13 @@ func (c *ApiController) UpdateOrganization() {
 		c.ResponseError(err.Error())
 		return
 	}
+	affected, err := object.UpdateOrganization(id, &organization)
+	if err != nil {
+		c.Data["json"] = wrapErrorResponse(err)
+	} else {
+		c.Data["json"] = wrapActionResponse(affected)
```
Comment (hsluoyz) [inline controllers/organization.go:86]: Add `wrapActionResponse2(affected, err)`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/organization.go [hunk 3]
```diff
@@ -79,8 +79,13 @@ func (c *ApiController) UpdateOrganization() {
 		c.ResponseError(err.Error())
```
Comment (hsluoyz) [inline controllers/organization.go:79]: All `panic(err)` in `organization.go` first, then for all files

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by yehong-z** [comment]
将organization中的panic改为了返回err的形式
影响的到controller层函数加上了
if err != nil {
    c.responseError()
}
影响到的其他model层暂时用以下形式改动
result, _ := func()
以保证代码能够运行。

File: controllers/account.go [hunk 4]
```diff
@@ -84,7 +84,12 @@ func (c *ApiController) Signup() {
 		return
 	}
 
-	organization := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", authForm.Organization))
+	organization, err := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", authForm.Organization))
```
Comment (hsluoyz) [inline controllers/account.go:87]: util.GetId

File: controllers/auth.go [hunk 5]
```diff
@@ -335,7 +339,11 @@ func (c *ApiController) Login() {
 			return
 		}
 
-		organization := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", application.Organization))
+		organization, err := object.GetOrganization(fmt.Sprintf("%s/%s", "admin", application.Organization))
```
Comment (hsluoyz) [inline controllers/auth.go:342]: util.GetId

File: object/organization.go [hunk 6]
```diff
@@ -68,56 +68,42 @@ type Organization struct {
 	AccountItems []*AccountItem `xorm:"varchar(3000)" json:"accountItems"`
 }
 
-func GetOrganizationCount(owner, field, value string) int {
+func GetOrganizationCount(owner, field, value string) (count int64, err error) {
```
Comment (hsluoyz) [inline object/organization.go:71]: int

File: object/organization.go [hunk 7]
```diff
@@ -68,56 +68,42 @@ type Organization struct {
...
-	return organizations
+	err = session.Find(&organizations)
+	return
 }
 
-func getOrganization(owner string, name string) *Organization {
+func getOrganization(owner string, name string) (*Organization, error) {
 	if owner == "" || name == "" {
-		return nil
+		return nil, fmt.Errorf("check:Organization does not exist")
```
Comment (hsluoyz) [inline object/organization.go:90]: owner or name is empty

File: object/organization.go [hunk 8]
```diff
@@ -68,56 +68,42 @@ type Organization struct {
...
+		return nil, err
 	}
 
 	if existed {
-		return &organization
+		return &organization, nil
 	}
 
-	return nil
+	return nil, fmt.Errorf("check:Organization does not exist")
```
Comment (hsluoyz) [inline object/organization.go:103]: Object does not exist

File: object/organization.go [hunk 9]
```diff
@@ -171,35 +157,27 @@ func UpdateOrganization(id string, organization *Organization) bool {
...
 
-func AddOrganization(organization *Organization) bool {
-	affected, err := adapter.Engine.Insert(organization)
-	if err != nil {
-		panic(err)
-	}
-
-	return affected != 0
+func AddOrganization(organization *Organization) (affected bool, err error) {
+	affectedRows, err := adapter.Engine.Insert(organization)
```
Comment (hsluoyz) [inline object/organization.go:167]: affected

File: object/organization.go [hunk 10]
```diff
@@ -171,35 +157,27 @@ func UpdateOrganization(id string, organization *Organization) bool {
...
+		return false, fmt.Errorf("Cannot delete built-in organization")
 	}
 
-	affected, err := adapter.Engine.ID(core.PK{organization.Owner, organization.Name}).Delete(&Organization{})
-	if err != nil {
-		panic(err)
-	}
-
-	return affected != 0
+	affectedRows, err := adapter.Engine.ID(core.PK{organization.Owner, organization.Name}).Delete(&Organization{})
```
Comment (hsluoyz) [inline object/organization.go:176]: affected

File: object/organization.go [hunk 11]
```diff
@@ -171,35 +157,27 @@ func UpdateOrganization(id string, organization *Organization) bool {
...
+func AddOrganization(organization *Organization) (affected bool, err error) {
+	affectedRows, err := adapter.Engine.Insert(organization)
+	return affectedRows != 0, err
 }
 
-func DeleteOrganization(organization *Organization) bool {
+func DeleteOrganization(organization *Organization) (affected bool, err error) {
 	if organization.Name == "built-in" {
-		return false
+		return false, fmt.Errorf("Cannot delete built-in organization")
```
Comment (hsluoyz) [inline object/organization.go:173]: prefix

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/account.go [hunk 12]
```diff
@@ -300,7 +305,13 @@ func (c *ApiController) GetAccount() {
 	user.Permissions = object.GetMaskedPermissions(user.Permissions)
 	user.Roles = object.GetMaskedRoles(user.Roles)
 
-	organization := object.GetMaskedOrganization(object.GetOrganizationByUser(user))
```
Comment (hsluoyz) [inline controllers/account.go:303]: What about `GetMaskedOrganization` accept 2nd arg of error?

File: controllers/account.go [hunk 13]
```diff
@@ -84,7 +84,12 @@ func (c *ApiController) Signup() {
 		return
 	}
 
-	organization := object.GetOrganization(util.GetId("admin", authForm.Organization))
+	organization, err := object.GetOrganization(util.GetId("admin", authForm.Organization))
+	if err != nil {
+		c.ResponseError(c.T(err.Error()))
```
Comment (hsluoyz) [inline controllers/account.go:89]: c.T will work for DB errors?

File: controllers/organization.go [hunk 14]
```diff
@@ -99,13 +128,23 @@ func (c *ApiController) AddOrganization() {
...
 
-	count := object.GetOrganizationCount("", "", "")
-	if err := checkQuotaForOrganization(count); err != nil {
+	count, err := object.GetOrganizationCount("", "", "")
+	if err != nil {
 		c.ResponseError(err.Error())
 		return
 	}
 
-	c.Data["json"] = wrapActionResponse(object.AddOrganization(&organization))
```
Comment (hsluoyz) [inline controllers/organization.go:108]: wrapActionResponse accepts new error arg

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: controllers/organization.go [hunk 15]
```diff
@@ -57,8 +74,7 @@ func (c *ApiController) GetOrganizations() {
 // @router /get-organization [get]
 func (c *ApiController) GetOrganization() {
 	id := c.Input().Get("id")
-
-	c.Data["json"] = object.GetMaskedOrganization(object.GetOrganization(id))
+	c.Data["json"], _ = object.GetMaskedOrganization(object.GetOrganization(id))
```
Comment (hsluoyz) [inline controllers/organization.go:77]: panic

File: ldap/util.go [hunk 16]
```diff
@@ -118,7 +118,7 @@ func GetFilteredUsers(m *ldap.Message) (filteredUsers []*object.User, code int)
 			return filteredUsers, ldap.LDAPResultSuccess
 		}
 
-		organization := object.GetOrganization(util.GetId("admin", org))
+		organization, _ := object.GetOrganization(util.GetId("admin", org))
```
Comment (hsluoyz) [inline ldap/util.go:121]: panic

File: ldap/util.go [hunk 17]
```diff
@@ -137,7 +137,7 @@ func GetFilteredUsers(m *ldap.Message) (filteredUsers []*object.User, code int)
 // TODO not handle salt yet
 // @return {md5}5f4dcc3b5aa765d61d8327deb882cf99
 func getUserPasswordWithType(user *object.User) string {
-	org := object.GetOrganizationByUser(user)
+	org, _ := object.GetOrganizationByUser(user)
```
Comment (hsluoyz) [inline ldap/util.go:140]: panic

File: object/application.go [hunk 18]
```diff
@@ -165,7 +165,7 @@ func extendApplicationWithProviders(application *Application) {
 }
 
 func extendApplicationWithOrg(application *Application) {
-	organization := getOrganization(application.Owner, application.Organization)
+	organization, _ := getOrganization(application.Owner, application.Organization)
```
Comment (hsluoyz) [inline object/application.go:168]: panic

File: object/check.go [hunk 19]
```diff
@@ -170,7 +170,7 @@ func CheckPassword(user *User, password string, lang string, options ...bool) st
 		}
 	}
 
-	organization := GetOrganizationByUser(user)
+	organization, _ := GetOrganizationByUser(user)
```
Comment (hsluoyz) [inline object/check.go:173]: panic

File: object/init.go [hunk 20]
```diff
@@ -74,7 +74,7 @@ func getBuiltInAccountItems() []*AccountItem {
 }
 
 func initBuiltInOrganization() bool {
-	organization := getOrganization("admin", "built-in")
+	organization, _ := getOrganization("admin", "built-in")
```
Comment (hsluoyz) [inline object/init.go:77]: panic

File: object/init_data.go [hunk 21]
```diff
@@ -174,7 +174,7 @@ func readInitDataFromFile(filePath string) *InitData {
 }
 
 func initDefinedOrganization(organization *Organization) {
-	existed := getOrganization(organization.Owner, organization.Name)
+	existed, _ := getOrganization(organization.Owner, organization.Name)
```
Comment (hsluoyz) [inline object/init_data.go:177]: panic

File: object/ldap_conn.go [hunk 22]
```diff
@@ -255,7 +255,7 @@ func SyncLdapUsers(owner string, syncUsers []LdapUser, ldapId string) (existUser
 		uuids = append(uuids, user.Uuid)
 	}
 
-	organization := getOrganization("admin", owner)
+	organization, _ := getOrganization("admin", owner)
```
Comment (hsluoyz) [inline object/ldap_conn.go:258]: panic

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**